cxx/emissions/categorical.hh

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include <cassert>
+#include <limits>
+#include <utility>
+
+#include "distributions/dirichlet_categorical.hh"
+#include "emissions/base.hh"
+
+// A "bigram" emission model that tracks separate emission distributions per
+// clean categorical state.
+class CategoricalEmission : public Emission<int> {
+ public:
+  mutable std::vector<DirichletCategorical> emission_dists;
+
+  CategoricalEmission(int num_states) {
+    emission_dists.reserve(num_states);
+    for (int i = 0; i < num_states; ++i) {
+      emission_dists.emplace_back(num_states);
+    }
+  };
+
+  void incorporate(const std::pair<int, int>& x) {
+    ++N;
+    emission_dists[x.first].incorporate(x.second);
+  }
+
+  void unincorporate(const std::pair<int, int>& x) {
+    --N;
+    emission_dists[x.first].unincorporate(x.second);
+  }
+
+  double logp(const std::pair<int, int>& x) const {
+    double lp;
+    for (size_t i = 0; i < emission_dists.size(); ++i) {
+      if (std::cmp_equal(i, x.first)) {
+        lp += emission_dists[i].logp(x.second);
+      } else {
+        lp += emission_dists[i].logp_score();
```

Comment:
Could you explain this formula for logp? I thought it should be `p(x.second | x.first)`, where `x.first` is clean and `x.second` is noisy, such that the conditional probability in the `if` branch was all we needed. Also, to the extent that this is like a bigram model, the bigram `logp` doesn't contain `logp_score` terms for the rest of the categories.

Reply:
You are entirely correct; this should only contain the conditional probability.  Fixed.